plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -251,5 +261,49 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName) : TypeEmbedding {
     override fun accessInvariants(v: Exp): List<Exp> =
         flatMapUniqueFields { _, field -> field.accessInvariantsForParameter(v) }
 
+    // Note: this function will replace accessInvariants when nested unfold will be implemented
+    override fun predicateAccessInvariants(v: Exp): List<Exp> = listOf(Exp.PredicateAccess(name, listOf(v)))
+
     override fun provenInvariants(v: Exp) = listOf(subtypeInvariant(v))
+
+    private val thisEmbedding = VariableEmbedding(ThisReceiverName, this)
+    val classPredicate: Predicate
+        get() {
+            val accessReadableFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.WildcardPerm()) }
+
+            val accessSpecialFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_WRITEABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.FullPerm()) }
+
+            val classFieldPredicates = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { it.type.predicateAccessInvariants(Exp.FieldAccess(Exp.LocalVar(ThisReceiverName, Type.Ref), it.toViper())) }
+                .flatten()
```

Comment:
Why not something like `fields.values.flatMap { it.someKindOfInvariants() }` or something like that?  It's a little unclear here what's going on.  (Perhaps only for the first two, not the last.)